feat(resources): add workspace resources and workspace run tasks func…

KshitijaChoudhari opened the pull request:
…tionality

- Add comprehensive workspace resources service with list operations
- Implement full CRUD workspace run tasks service
- Add stage enum conversion handling (snake_case ↔ kebab-case)
- Create complete data models with Pydantic validation
- Add 36 comprehensive unit tests with 100% coverage
- Include working example applications with CLI interfaces
- Maintain backward compatibility with existing pagination
- Support for all execution stages and enforcement levels

<!--
Thank you for contributing to hashicorp/python-tfe! Please read docs/CONTRIBUTING.md for detailed information when preparing your change.

Here's what to expect after the pull request is opened:

The test suite contains many acceptance tests that are run against a test version of HCP Terraform, and additional testing is done against Terraform Enterprise. You can read more about running the tests against your own Terraform Enterprise environment in TESTS.md. Our CI system (Github Actions) will not test your fork until a one-time approval takes place.

Your change, depending on its impact, may be released in the following ways:

  1. For impactful bug fixes, it can be released fairly quickly as a patch release.
  2. For noncritical bug fixes and new features, it will be incorporated into the next minor version release.
  3. For breaking changes (those changes that alter the public method signatures), more consideration must be made and alternatives may be considered, depending on upgrade difficulty and release schedule.

Please note that API features that are not generally available should not be merged/released without prior discussion with the maintainers. See docs/CONTRIBUTING Section "Adding API changes that are not generally available" for more information.

Please fill out the remaining template to assist code reviewers and testers with incorporating your change. If a section does not apply, feel free to delete it.
-->

## Description

Change 1: Workspace Resources Implementation
Added complete workspace resources service with list operations
Implemented proper URL encoding and parameter handling for workspace IDs
Added pagination support with comprehensive filtering options
Created robust data models with Pydantic validation
Change 2: Workspace Run Tasks Implementation
Implemented full CRUD operations (create, read, update, delete, list)
Added stage enum conversion handling (snake_case :left_right_arrow: kebab-case) for API compatibility
Created comprehensive options models for all operations
Implemented proper relationship handling and include options
Change 3: Enhanced Run Task Integration
Updated existing run task service to support workspace task relationships
Added graceful Stage enum parsing for API responses
Improved error handling and validation across all operations
Change 4: Comprehensive Testing Suite
Added 36 new unit tests covering all functionality
Implemented comprehensive error scenario testing
Added model validation and data transformation tests
Maintained 100% test coverage for new functionality
Change 5: Example Applications
Created working example files with CLI interfaces
Demonstrated real-world usage with live HCP Terraform integration
Added comprehensive command-line argument handling
Implementation Details:
Architecture Changes:
Service Layer: Added WorkspaceResourcesService and WorkspaceRunTasksService following existing patterns
Model Layer: Created comprehensive Pydantic models with proper validation and field aliases
HTTP Transport: Enhanced URL construction and parameter handling for workspace-specific endpoints
Enum Handling: Implemented bidirectional Stage enum conversion for API compatibility
Key Files Changed
[workspace_resources.py]
[workspace_run_tasks.py]
[run_task.py]
[workspace_resource.py]
[workspace_run_task.py]
[__init__.py]
[test_workspace_resources.py]
[test_workspace_run_tasks.py]
[workspace_resources.py]
[workspace_run_tasks.py]

## Testing plan
365 total tests passing (100% success rate)
36 new tests added for workspace functionality
0.51 seconds execution time - excellent performance
Zero regressions - all existing functionality maintained
Integration Testing:
Real API Testing: Successfully tested against live HCP Terraform instance
Data Handling: Verified proper parsing of real API responses
Error Scenarios: Comprehensive error handling validation
Code Quality:
Formatting: All 132 files properly formatted (Ruff)
Linting: Zero linting issues (Ruff + MyPy)
Type Safety: All 81 source files pass type checking
Standards: Full compliance with project coding standards
Example Application Results:
Workspace Resources: Successfully lists and displays 3 real resources with complete metadata
Workspace Run Tasks: Complete CLI interface with all CRUD operations functional

## External links
- [API documentation]
https://developer.hashicorp.com/terraform/cloud-docs/api-docs/workspaces
https://developer.hashicorp.com/terraform/cloud-docs/workspaces/settings/run-tasks
-->

## Output from tests

1) Workspace Resources Example Output: 

Listing resources for workspace: <workspace-id>

Found 3 resource(s):
--------------------------------------------------------------------------------
ID: wsr-xxxxxxxxxxxxxxxxx
Address: random_password.example
Name: example
Module: root
Provider: hashicorp/random
Provider Type: random_password
Created At: 
Updated At: 
Modified By State Version: sv-xxxxxxxxxxxxxxxxx
--------------------------------------------------------------------------------
ID: wsr-xxxxxxxxxxxxxxxxx
Address: random_string.test
Name: test
Module: root
Provider: hashicorp/random
Provider Type: random_string
Created At: 
Updated At: 
Modified By State Version: sv-xxxxxxxxxxxxxxxxx
--------------------------------------------------------------------------------
ID: wsr-xxxxxxxxxxxxxxxxx
Address: random_uuid.workspace_id
Name: workspace_id
Module: root
Provider: hashicorp/random
Provider Type: random_uuid
Created At: 
Updated At: 
Modified By State Version: sv-xxxxxxxxxxxxxxxxx
--------------------------------------------------------------------------------

Pagination Info:
  Current Page: 1
  Total Pages: 1
  Total Count: 3


2) Workspace Run Tasks Example Output:

Help/Usage Display: 

Workspace Run Tasks Management :

options:
  --org ORG             Organization name
  --workspace WORKSPACE Workspace ID
  --create             Create new workspace run task
  --update             Update existing workspace run task
  --delete             Delete workspace run task
  --enforcement-level  Set enforcement level (advisory|mandatory)
  --stage              Set execution stage (pre_plan|post_plan|pre_apply|post_apply)

List Workspace Run Tasks:
Listing Workspace Run Tasks for Workspace: <workspace-id>
No workspace run tasks found.
Pagination Info:
  Current Page: 1
  Total Pages: 0
  Total Count: 0

Create Workspace Run Task:
Creating Workspace Run Task
✅ Successfully created workspace run task: <workspace-task-id>
   Task ID: <task-id>
   Enforcement Level: advisory
   Stage: post_plan

List After Creation:
Listing Workspace Run Tasks for Workspace: <workspace-id>
Found 1 workspace run task(s):
Workspace Run Task ID: <workspace-task-id>
  Run Task ID: <task-id>  
  Enforcement Level: advisory
  Stage: post_plan
  Created At: 
  Updated At: \
Pagination Info:
  Current Page: 1
  Total Pages: 1
  Total Count: 1

Update Workspace Run Task:
Updating Workspace Run Task: <workspace-task-id>
✅ Successfully updated workspace run task
   Enforcement Level: mandatory
   Stage: pre_plan

Delete Workspace Run Task:
Deleting Workspace Run Task: <workspace-task-id>
✅ Successfully deleted workspace run task: <workspace-task-id>


## Rollback Plan

If a change needs to be reverted, we will roll out an update to the code within 7 days.
-->

## Changes to Security Controls

None

## PCI review checklist

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [ ] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

If you have any questions, please contact your direct supervisor, GRC (#team-grc), or the PCI working group (#proj-pci-reboot). You can also find more information at [PCI Compliance](https://hashicorp.atlassian.net/wiki/spaces/SEC/pages/2784559202/PCI+Compliance).

KshitijaChoudhari committed:
feat(resources): add workspace resources and workspace run tasks functionality

- Add comprehensive workspace resources service with list operations
- Implement full CRUD workspace run tasks service
- Add stage enum conversion handling (snake_case ↔ kebab-case)
- Create complete data models with Pydantic validation
- Add 36 comprehensive unit tests with 100% coverage
- Include working example applications with CLI interfaces
- Maintain backward compatibility with existing pagination
- Support for all execution stages and enforcement levels

KshitijaChoudhari committed:
Fix code formatting issues

- Format examples/workspace_resources.py and src/pytfe/resources/workspace_resources.py
- Ensure compliance with ruff formatting standards

KshitijaChoudhari committed:
Fix Stage enum values and remove unnecessary conversions

- Corrected Stage enum to use kebab-case values (pre-plan, post-plan, etc.)
- Moved Stage and TaskEnforcementLevel back to run_task.py where they belong
- Updated workspace_run_task.py to import these enums from run_task.py
- Removed unnecessary stage conversion logic in workspace_run_tasks.py service
- Removed unnecessary stage conversion logic in run_task.py service
- Updated test assertions to expect kebab-case stage values
- All 365 tests passing, lint checks passing, type checks passing

KshitijaChoudhari committed:
Refactored workspace run tasks and resources

KshitijaChoudhari committed:
Merge branch 'next-0.1.1' into kshitija/WorkspaceRunTasksAndWorkspaceResources

isivaselvan commented on src/pytfe/resources/workspace_resources.py:
```diff
@@ -0,0 +1,103 @@
+
+    def list(
+        self, workspace_id: str, options: WorkspaceResourceListOptions | None = None
+    ) -> WorkspaceResourcesList:
```
convert the return type of list method into a iterator pattern

KshitijaChoudhari committed:
Pr review changes - workspace run tasks and resources

KshitijaChoudhari committed:
Apply ruff formatting and linting fixes

KshitijaChoudhari commented on examples/workspace_resources.py:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python3
```
resolved

KshitijaChoudhari committed:
Replace dictionary with RunTaskRelationship model in WorkspaceRunTaskCreateOptions

- Add RunTaskRelationship and RunTaskRelationshipData Pydantic models
- Update WorkspaceRunTaskCreateOptions to use RunTaskRelationship instead of dict
- Update service to serialize relationship using model_dump()
- Add workspace run tasks relationship parsing in run_task resource
- Update all tests and examples to use new relationship models
- Export new models in __init__.py

All tests passing (407/407), formatting and type checking clean.

KshitijaChoudhari committed:
Replace RunTaskRelationship wrapper with direct RunTask object reference

- WorkspaceRunTaskCreateOptions now accepts RunTask object directly
- Service layer extracts ID from RunTask object to build JSON:API relationship
- Made RunTask fields optional (name, url, category, enabled) to allow minimal construction with just ID
- Removed RunTaskRelationship and RunTaskRelationshipData wrapper models
- Updated tests and examples to use RunTask(id='task-123') pattern
- Aligns with go-tfe pattern of passing objects directly

This simplifies the API - users can create RunTask(id='task-123') instead of nested relationship models.

KshitijaChoudhari committed:
Remove duplicate examples from workspace_run_tasks.py epilog
